taichi/program/state_flow_graph.h

```diff
@@ -31,16 +31,25 @@ class StateFlowGraph {
     TaskMeta *meta{nullptr};
     // Incremental ID to identify the i-th launch of the task.
     bool is_initial_node{false};
+    bool executed{false};
```

Comment:
nit: I think `executed` and `pending_node_id == -1` mean the same thing, which means this state is not normalized in one place. How about this:

```cpp
bool executed() const { return pending_node_id == -1; }
void mark_executed() { pending_node_id = -1; }
```